### PR TITLE
Fix rare case of CHECKSUM_DOESNT_MATCH error

### PR DESCRIPTION
**Pull request by alesapin**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix rare bug in replicated merge tree which could lead to self-recovering `CHECKSUM_DOESNT_MATCH` error in logs.

### CODE REVIEW

File: src/Storages/MergeTree/MergeTreeData.cpp [hunk 1]
```diff
@@ -4245,7 +4245,7 @@ void MergeTreeData::forcefullyMovePartToDetachedAndRemoveFromMemory(const MergeT
 }
 
 
-void MergeTreeData::tryRemovePartImmediately(DataPartPtr && part)
+bool MergeTreeData::tryRemovePartImmediately(DataPartPtr && part)
```
Comment (alesapin) [inline src/Storages/MergeTree/MergeTreeData.cpp:4248]: Unrelated to changes

File: src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp [hunk 2]
```diff
@@ -801,8 +802,46 @@ std::pair<std::vector<String>, bool> ReplicatedMergeTreeSinkImpl<async_insert>::
...
+                {
+                    storage.checkPartChecksumsAndAddCommitOps(zookeeper, part, ops, existing_part_name);
+                }
+                catch (const zkutil::KeeperException &)
+                {
+                    throw;
+                }
+                catch (const Exception & ex)
+                {
+                    if (ex.code() == ErrorCodes::CHECKSUM_DOESNT_MATCH)
```
Comment (alesapin) [inline src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp:820]: Fix is here, don't commit parts when checksums are actually different.

**Comment by robot-ch-test-poll** [comment]
This is an automated comment for commit f3729da64ba4c1ece9a14318c264a9958ae6805a with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/54549/f3729da64ba4c1ece9a14318c264a9958ae6805a/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Review by CheSema — APPROVED** [review]

File: tests/queries/0_stateless/02882_replicated_fetch_checksums_doesnt_match.sql [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+INSERT INTO checksums_r3 VALUES (1, 'hello');
+
+SYSTEM START REPLICATION QUEUES checksums_r2;
+
+SYSTEM SYNC REPLICA checksums_r2;
+
+SELECT count() FROM checksums_r1;
+SELECT count() FROM checksums_r2;
+SELECT count() FROM checksums_r3;
```
Comment (vdimir) [inline tests/queries/0_stateless/02882_replicated_fetch_checksums_doesnt_match.sql:26]: So, it will be possible that we don't see data on a particular replica right after inserting to this replica, until part with the expected checksum is fetched, right?

Comment (alesapin) [reply tests/queries/0_stateless/02882_replicated_fetch_checksums_doesnt_match.sql:26]: Right, and I think it's more expected behavior.

**Comment by alesapin** [comment]
Failures unrelated.

File: src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp [hunk 4]
```diff
@@ -3,6 +3,7 @@
 #include <Storages/MergeTree/ReplicatedMergeTreeSink.h>
 #include <Storages/MergeTree/InsertBlockInfo.h>
 #include <Interpreters/PartLog.h>
+#include "Common/Exception.h"
```
Comment (azat) [inline src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp:6]: Apparently you were using `clangd`...

File: src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp [hunk 5]
```diff
@@ -801,8 +803,48 @@ std::pair<std::vector<String>, bool> ReplicatedMergeTreeSinkImpl<async_insert>::
...
+                    /// if we found part in deduplication hashes part must exists on some replica
+                    storage.checkPartChecksumsAndAddCommitOps(zookeeper, part, ops, existing_part_name, unused);
+                }
+                catch (const zkutil::KeeperException &)
+                {
+                    throw;
+                }
+                catch (const Exception & ex)
+                {
+                    if (ex.code() == ErrorCodes::CHECKSUM_DOESNT_MATCH)
```
Comment (tavplubix) [inline src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp:823]: Note that it still will show `CHECKSUM_DOESNT_MATCH` errors in `system.errors`. It would be much better to introduce a function overload that returns bool instead of throwing exception
